### PR TITLE
Enable BLEU/TER (& custom) scoring during training and validation

### PR DESCRIPTION
**Pull request by l-k-11235**
#1158 #2166

This PR offers to dynamcially calculate the BLEU and TER scores of training and validation in order to visualize them on tensorboard as well as the accuracy.
Thanks to a new  `scorers` module  inspired from `transforms`, it can be easily customized to support new metrics.\
Several flags have been added.\
- To calculate BLEU and TER of training:\
`--train_metrics [BLEU, TER]`
- To perform it every 200 steps:
`--train_eval_steps 200`
- To calculate BLEU and TER of validation:\
`--eval_metrics [BLEU, TER]`
- To dump the batch predictions involved in those calculations in fhe folder `dump_preds`:\
`--scoring_debug --dump_preds dump_preds`

This extends on the idea of #2161. Thanks @kagrze for the suggestion and initial work!

### CODE REVIEW

**Comment by vince62s** [comment]
Great! would that make sense to make report_every the same train_eval_steps ? if there is no overheads in computing those metrics I think we can make it the same.

**Comment by francoishernandez** [comment]
There is some significant overhead, since translation is performed, hence the decoupling of both values.

**Review by francoishernandez — APPROVED** [review]

**Comment by anderleich** [comment]
Hi @francoishernandez ,
I was wondering how this new functionality works? Are the source sentences being translated with beam search for the BLEU/TER computation for training and validation, or just during validation? Is it really necessary to perform the translation, couldn't we just use the model outputs?

**Comment by aeceou** [comment]
I'm sorry, but I don't understand why the statement "valid_stats = self._maybe_gather_stats(valid_stats)" has been removed from trainer.py.
As far as I know, the loss from each device must be added up at that stage so that we know the cross-entropy and perplexity.
Plus, shouldn't the corpus scores be calculated over the whole corpus, not over the batch?

**Comment by francoishernandez** [comment]
@anderleich 
> Hi @francoishernandez , I was wondering how this new functionality works? Are the source sentences being translated with beam search for the BLEU/TER computation for training and validation, or just during validation? Is it really necessary to perform the translation, couldn't we just use the model outputs?

Check the code: translation is performed. This is necessary, as stated in various previous discussions. Model outputs are not a generated translation, just the prediction of the next word.
As for training/validation, both are configurable via the new flags.

@aeceou 
> I'm sorry, but I don't understand why the statement "valid_stats = self._maybe_gather_stats(valid_stats)" has been removed from trainer.py.
As far as I know, the loss from each device must be added up at that stage so that we know the cross-entropy and perplexity.
Plus, shouldn't the corpus-scores be calculated over the whole corpus, not over a batch?

Indeed. The valid computation was supposedly restricted to a single gpu, but seems like we lost track of some changes. Investigating. 
The whole corpus is retrieved from the batch via `batch.dataset`. A bit of a workaround but seems cleaner than explicitly reading the files from disk again.

**Comment by francoishernandez** [comment]
@aeceou https://github.com/OpenNMT/OpenNMT-py/pull/2205 will make things clearer.
Thanks for raising the issue.

**Comment by aeceou** [comment]
> @aeceou #2205 will make things clearer. Thanks for raising the issue.

@francoishernandez
Thanks for your reply.
My second question was answered.
However, the loss would not be calculated over the whole data set (by refering to `Batch.dataset`); each device acculmulates its own loss, as far as I know.
Thus, if we are talking about BLEU or TER, which involve 'translation', yes, I agree with you, but I'm still not sure if we can really get the right cross-entropy and perplexity if we do not gather the loss values.

**Comment by aeceou** [comment]
> > @aeceou #2205 will make things clearer. Thanks for raising the issue.
> 
> @francoishernandez Thanks for your reply. My second question was answered. However, the loss would not be calculated over the whole data set (by refering to `Batch.dataset`); each device acculmulates its own loss, as far as I know. Thus, if we are talking about BLEU or TER, which involve 'translation', yes, I agree with you, but I'm still not sure if we can really get the right cross-entropy and perplexity if we do not gather the loss values.

@francoishernandez
Now I get the idea.
`valid_iter = IterOnDevice(valid_iter, device_id)` lets each process read over the whole corpus and thus removes the need to gather the loss values.
Must have been just an obsolete statement left there.
Thanks😊
